src/common/async/co_throttle.h

```diff
@@ -0,0 +1,113 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 Red Hat <contact@redhat.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <limits>
+#include <boost/intrusive_ptr.hpp>
+#include "common/async/cancel_on_error.h"
+#include "common/async/detail/co_throttle_impl.h"
+
+namespace ceph::async {
+
+/// A coroutine throttle that allows a parent coroutine to spawn and manage
+/// multiple child coroutines, while enforcing an upper bound on concurrency.
+///
+/// Child coroutines must be of type awaitable<void>. Exceptions thrown by
+/// children are rethrown to the parent on its next call to spawn() or wait().
+/// The cancel_on_error option controls whether these exceptions errors trigger
+/// the cancellation of other children.
+///
+/// All child coroutines are canceled by cancel() or co_throttle destruction.
+/// This allows the parent coroutine to share memory with its child coroutines
+/// without fear of dangling references.
+///
+/// This class is not thread-safe, so a strand executor should be used in
+/// multi-threaded contexts.
+///
+/// Example:
+/// \code
+/// awaitable<void> child(task& t);
+///
+/// awaitable<void> parent(std::span<task> tasks)
+/// {
+///   // process all tasks, up to 10 at a time
+///   auto ex = co_await boost::asio::this_coro::executor;
+///   auto throttle = co_throttle{ex, 10};
+///
+///   for (auto& t : tasks) {
+///     co_await throttle.spawn(child(t));
+///   }
+///   co_await throttle.wait();
+/// }
+/// \endcode
+template <boost::asio::execution::executor Executor>
+class co_throttle {
```

Comment:
I'm not sure how much I like none of these being thread safe. I understand we don't want the overhead when we don't need to pay for it, but the lack of thread safety makes some things a bit difficult. I couldn't figure out how to get `co_throttle` to work with datalog recovery in a way that didn't crash on cancellation, for example/

Could we maybe have a template parameter to determine it?

Reply:
i'm not sure where the gap is with cancellation. i added a unit test for both co_throttle and co_spawn_group that trigger cancellation from a different thread, and neither thread-sanitizer nor valgrind complain about data races. can you suggest any changes to that test to make it more like the datalog case?

Reply:
when designing this stuff, i admittedly had a very specific use case in mind: allowing a parent coroutine to spawn and manage several child coroutines in a way that guarantees they can safely access memory on the parent's stack

as long as the co_throttle/spawn_group is owned by the parent stack, cancellation of the parent will cause the stack to unwind and the co_foo to be destroyed, which cancels any child coroutines before they have another chance to access the parent stack

here it's crucial that the parent coroutine is running on the same executor as the children so that this child cancellation is synchronous. if they're on different executors, the child cancellations will get dispatched asynchronously, leaving a window where they can race to access the parent stack that's gone away

i can certainly understand the desire to generalize from 'parent coroutine' to 'parent class', but it's very hard to guarantee that a class destructor will run on a specific executor. in the case of Datalog, rgw's shutdown is likely running outside of our io_context_pool entirely

in terms of thread safety, it should be possible for the class destructor to safely trigger cancellation of the children. but we can't guarantee that the child coroutines can safely access member variables or call member functions/coroutines which is a major limitation

Reply:
the `co_throttle` interface does a better job of enforcing this guarantee, because `co_throttle::spawn()` wraps the call to `asio::co_spawn()` so always passes the correct executor

`co_spawn_group` should do the same, instead of functioning as a `CompletionToken` for manual calls to `asio::co_spawn()`. i'd thought that was a more flexible design, but not at the cost of safety